### PR TITLE
docs: Enhance developer documentation for testing and static analysis

### LINKED ISSUE #2279: Improve Public Facing Developer Documentation on System Testing
Mantid Imaging's (MI) externally facing developer documentation surrounding system testing does not go into any particular depth which could be seen as bot being inclusive of introducing the larger open source community to the repository or new starters within the team.

We should improve documentation surrounding usage of MI's tests and checks accessible within our `Makefile` and how to modify them for certain circumstances such as adding delays to screenshot tests. 

Plan, discuss and finally update MI's developer documentation surrounding system testing and static analysis checks to be more inclusive for users who are new to MI.

Examples of commands within Mantid Imaging's `makefile` which should be documented more including how to modify them to achieve certain results are listed below:
* `test`
* `test-verbose`
* `test-system`
* `test-screenshots`
* `test-screenshots-win`
* `mypy`
* `yapf`
* `ruff`
* `check`

### PR DESCRIPTION
**Pull request by ashmeigh**
### Issue

Closes #2279 

### Description

This PR enhances the public-facing developer documentation for Mantid Imaging, making it more accessible to new contributors. Key updates include:

### Testing 

This PR enhances the public-facing developer documentation for Mantid Imaging, making it more accessible to new contributors. Key updates include:

Detailed Descriptions of make Commands:
Expanded explanations for commands like make test, make test-verbose, make test-system, and more.
Instructions on modifying commands, such as adding delays to screenshot tests.

Static Analysis Tools:
Added guidance on using mypy, ruff, and yapf.

Pre-Commit Hooks:
Instructions for setting up and using pre-commit hooks.

### Acceptance Criteria 

Confirm the documentation's clarity and correctness.
Verify the functionality of the make commands as described.

### Documentation

Changes have been added to the appropriate sections, including docs/release_notes, for the upcoming release.

### CODE REVIEW

**Comment by ashmeigh** [comment]
Other things that i think should be added are 

Running a Specific Test Module:
python -m unittest 

Using pytest for Specific Tests:
pytest -k "test"

More detail in the logging and GUI

File: docs/developer_guide/testing.rst [hunk 1]
```diff
@@ -38,22 +67,52 @@ For options such as running a subset of tests, see `PyTest Docs <https://docs.py
 Static analysis
 ---------------
 
-Mantid Imaging uses `mypy <http://mypy-lang.org/>`_, `ruff <https://beta.ruff.rs/docs/>`_ and `yapf <https://github.com/google/yapf>`_ for static analysis and formatting. They are run by :code:`make check`, or can be run individually, e.g. :code:`make mypy`.
+Mantid Imaging uses several tools to ensure high code quality and consistency:
+
+- :code:`mypy <http://mypy-lang.org/>`: Enforces type correctness to catch potential type errors early.
+- :code:`ruff <https://beta.ruff.rs/docs/>`: Checks for common syntax and style issues to maintain clean, readable code.
+- :code:`yapf <https://github.com/google/yapf>`: Formats code according to PEP 8 standards, ensuring consistent style.
```
Comment (samtygier-stfc) [inline docs/developer_guide/testing.rst:74]: Can you make these links, e.g.

```
- `mypy <http://mypy-lang.org/>`_: Enforces type correctness to catch potential type errors early.
```

File: docs/developer_guide/testing.rst [hunk 2]
```diff
@@ -1,29 +1,58 @@
...
+Refer to the :file:`Makefile` for additional options or to modify these commands for specific scenarios.
+
+Running Specific Tests
+----------------------
+
+Mantid Imaging allows you to focus on specific test cases or workflows by running individual test modules or selecting tests based on keywords.
+
+To run a specific test module, you can specify its path using the following command:
+
+    python -m unittest path.to.your_test_module
```
Comment (samtygier-stfc) [inline docs/developer_guide/testing.rst:31]: Need a double colon `::` to make a code box.

Also, lets stick to pytest commands for consistancy

`pytest path/to/test_file.py`

File: docs/developer_guide/testing.rst [hunk 3]
```diff
@@ -38,22 +67,52 @@ For options such as running a subset of tests, see `PyTest Docs <https://docs.py
...
-Mantid Imaging uses `Applitools Eyes <https://applitools.com/products-eyes/>`_ for GUI approval testing. Screenshots of windows are uploaded and compared to known good baseline images. This is run in the github action on pull requests.
+Mantid Imaging uses `Applitools Eyes <https://applitools.com/products-eyes/>`_ for GUI approval testing. During this process, screenshots of application windows are captured and compared against known good baseline images to detect any unintended visual changes. This process is integrated into the GitHub Actions workflow and runs automatically on pull requests.
+
+The tests can also be run locally, with options tailored for different platforms:
+
+- **Cross-Platform Testing**: Run the tests on any platform and save the screenshots to a specified directory:
+
+      make test-screenshots
+
+  By default, screenshots are saved to a predefined directory. You can customize the output directory by setting the :code:`APPLITOOLS_IMAGE_DIR` environment variable. For example:
```
Comment (samtygier-stfc) [inline docs/developer_guide/testing.rst:98]: `::`

File: docs/developer_guide/testing.rst [hunk 4]
```diff
@@ -38,22 +67,52 @@ For options such as running a subset of tests, see `PyTest Docs <https://docs.py
...
 
-The tests can be run locally, where the screenshots will be writen to a directory::
+      export APPLITOOLS_IMAGE_DIR=/path/to/your/directory
 
-    make test-screenshots
+- **Windows-Specific Testing**: For running screenshot tests specifically on Windows, use the following command:
```
Comment (samtygier-stfc) [inline docs/developer_guide/testing.rst:102]: `::`

File: docs/developer_guide/testing.rst [hunk 5]
```diff
@@ -38,22 +67,52 @@ For options such as running a subset of tests, see `PyTest Docs <https://docs.py
...
 
-See the :file:`Makefile` for the underlying command used. You may want to run it with a specific output directory by changing :code:`APPLITOOLS_IMAGE_DIR`, or to upload to the Applitools API by setting a :code:`APPLITOOLS_API_KEY` (can be found in the Applitools web interface).
+      make test-screenshots-win
 
+  This command ensures that the graphical output is consistent on Windows, accounting for any platform-specific rendering differences.
 
-When these tests are run from the automated tests differences between uploaded and baseline images can be examined and approved or rejected from the Applitools web interface.
+If you want to upload the screenshots to the Applitools API for comparison with baseline images stored on their servers, you need to set the :code:`APPLITOOLS_API_KEY` environment variable, which can be obtained from the Applitools web interface:
```
Comment (samtygier-stfc) [inline docs/developer_guide/testing.rst:108]: `::`

File: docs/developer_guide/testing.rst [hunk 6]
```diff
@@ -1,29 +1,58 @@
...
+
+To run a specific test module, you can specify its path using the following command:
+
+    python -m unittest path.to.your_test_module
+
+If you are using `pytest`, you can run tests that match a specific keyword by using the ``-k`` flag::
+
+    pytest -k "test_valid_login"
+
+These commands allow you to target specific areas of the codebase during development, making it easier to test and debug individual components or workflows.
```
Comment (samtygier-stfc) [inline docs/developer_guide/testing.rst:38]: Can we add:
```
See the Pytest docs for additional `command line options <https://docs.pytest.org/en/stable/reference/reference.html#command-line-flags>`.
```

**Review by samtygier-stfc — CHANGES_REQUESTED** [review]

File: docs/developer_guide/testing.rst [hunk 7]
```diff
@@ -1,34 +1,63 @@
...
+Refer to the :file:`Makefile` for additional options or to modify these commands for specific scenarios.
+
+Running Specific Tests
+----------------------
+
+Mantid Imaging allows you to focus on specific test cases or workflows by running individual test modules or selecting tests based on keywords.
+
+To run a specific test module, you can specify its path using the following command::
+
+    pytest -m path/to/test_file.py
```
Comment (samtygier-stfc) [inline docs/developer_guide/testing.rst:31]: I think the `-m` is not need here.

File: docs/developer_guide/testing.rst [hunk 8]
```diff
@@ -1,34 +1,63 @@
...
+
+If you are using `pytest`, you can run tests that match a specific keyword by using the ``-k`` flag::
+
+    pytest -k "test_valid_login"
+
+These commands allow you to target specific areas of the codebase during development, making it easier to test and debug individual components or workflows.
+See the Pytest docs for additional `command line options <https://docs.pytest.org/en/stable/reference/reference.html#command-line-flags>`.
```
Comment (samtygier-stfc) [inline docs/developer_guide/testing.rst:38]: Needs the `_` at the end to make it a link.

```
See the Pytest docs for additional `command line options <https://docs.pytest.org/en/stable/reference/reference.html#command-line-flags>`_.
```

File: docs/developer_guide/testing.rst [hunk 9]
```diff
@@ -38,22 +67,52 @@ For options such as running a subset of tests, see `PyTest Docs <https://docs.py
 Static analysis
 ---------------
 
-Mantid Imaging uses `mypy <http://mypy-lang.org/>`_, `ruff <https://beta.ruff.rs/docs/>`_ and `yapf <https://github.com/google/yapf>`_ for static analysis and formatting. They are run by :code:`make check`, or can be run individually, e.g. :code:`make mypy`.
+Mantid Imaging uses several tools to ensure high code quality and consistency:
+
+- :code:`mypy <http://mypy-lang.org/>`_: Enforces type correctness to catch potential type errors early.
+- :code:`ruff <https://beta.ruff.rs/docs/>`_: Checks for common syntax and style issues to maintain clean, readable code.
+- :code:`yapf <https://github.com/google/yapf>`_: Formats code according to PEP 8 standards, ensuring consistent style.
+
```
Comment (samtygier-stfc) [inline docs/developer_guide/testing.rst:75]: These need the `:code:` removing to make the links work

**Review by samtygier-stfc — CHANGES_REQUESTED** [review]

**Review by samtygier-stfc — APPROVED** [review]
Thanks. Links all work now.

**Pull request merged**